examples/07_tutorials/ai-reco-of-recos/utils/openai_utils.py

```diff
@@ -0,0 +1,456 @@
+    return result.choices[0].message.content
+      
+@retry(wait=wait_random_exponential(min=1, max=30), stop=stop_after_attempt(10))
+def call_o1(messages,  client, model, reasoning_effort ="medium"): 
+    # print(f"\nCalling OpenAI APIs with {len(messages)} messages - Model: {model} - Endpoint: {client._base_url}\n")
+    response = client.chat.completions.create(model=model, messages=messages, reasoning_effort=reasoning_effort)
+    return response.model_dump()['choices'][0]['message']['content']
+
+@retry(wait=wait_random_exponential(min=1, max=30), stop=stop_after_attempt(10))
+def call_o1_mini(messages,  client, model): 
+    # print(f"\nCalling OpenAI APIs with {len(messages)} messages - Model: {model} - Endpoint: {client._base_url}\n")
+    response = client.chat.completions.create(model=model, messages=messages)
+    return response.model_dump()['choices'][0]['message']['content']
+
+@retry(wait=wait_random_exponential(min=1, max=30), stop=stop_after_attempt(10))       
+def call_o3(messages,  client, model, reasoning_effort ="medium"): 
+    print(f"\ncall_o3:: Calling OpenAI APIs with {len(messages)} messages - Model: {model} - Endpoint: {client._base_url}\n")
+    response = client.chat.completions.create(model=model, messages=messages, reasoning_effort=reasoning_effort)
+    return response.model_dump()['choices'][0]['message']['content']
+
+@retry(wait=wait_random_exponential(min=1, max=30), stop=stop_after_attempt(10))
+def call_o3_mini(messages,  client, model, reasoning_effort ="medium"): 
+    # print(f"\nCalling OpenAI APIs with {len(messages)} messages - Model: {model} - Endpoint: {client._base_url} - Reasoning Effort: {reasoning_effort}\n")
+    response = client.chat.completions.create(model=model, messages=messages, reasoning_effort=reasoning_effort)
+    return response.model_dump()['choices'][0]['message']['content']
+
+@retry(wait=wait_random_exponential(min=1, max=30), stop=stop_after_attempt(10))
+def call_o4_mini(messages, client, model, reasoning_effort ="medium"): 
+    print(f"\ncall_o4_mini::Calling OpenAI APIs with {len(messages)} messages - Model: {model} - Endpoint: {client._base_url}\n")
+    response = client.chat.completions.create(model=model, messages=messages, reasoning_effort=reasoning_effort)
+    return response.model_dump()['choices'][0]['message']['content']
+
+
+def call_llm_structured_outputs(prompt: str, model_info: Union[MulitmodalProcessingModelInfo, TextProcessingModelnfo], response_format, imgs=[]):
+    content = [{"type": "text", "text": prompt}]
+    content = content + prepare_image_messages(imgs)
+    messages = [
+        {"role": "user", "content": content},
+    ]
+
+    if model_info.client is None: model_info = instantiate_model(model_info)
+    # print(">>>>>>>>>>>>>>>>> call_llm_structured_outputs model_info", model_info)
+
+
+    if (model_info.model_name == "gpt-4o") or ((model_info.model_name == "gpt-45")):
```

Comment:
same here, if you use Responses API, this conditional code will get much simpler.